lib/import-map.php

```diff
@@ -0,0 +1,133 @@
+gutenberg_register_import_map_entry( 'moment-timezone/moment-timezone-utils', 'https://cdn.skypack.dev/moment-timezone/moment-timezone-utils' );
+
+/**
+ * Convert string to in camel-case, useful for class name patterns.
+ *
+ * @param string $string Target string.
+ *
+ * @return string Camel-case string.
+ */
+function gutenberg_to_camel_case( $string ) {
+	$string = str_replace( '-', ' ', $string );
+	$string = str_replace( '_', ' ', $string );
+	$string = lcfirst( ucwords( strtolower( $string ) ) );
+	$string = str_replace( ' ', '', $string );
+	return $string;
+}
+
+foreach ( glob( gutenberg_dir_path() . 'build/esm/*.js' ) as $path ) {
+	$package_name = substr( basename( $path ), 0, -3 );
+	gutenberg_register_import_map_entry( '@wordpress/' . $package_name, plugins_url( 'build/esm', __DIR__ ) . '/' . $package_name . '.js' );
+}
+
+/**
+ * Hichjack the scripts initialization to use the esmodules instead.
+ *
+ * @param string $tag    The `<script>` tag for the enqueued script.
+ * @param string $handle The script's registered handle.
+ * @param string $src    The script's source URL.
+ *
+ * @return string Updated tag.
+ */
+function gutenberg_inject_module_scripts( $tag, $handle, $src ) {
+	$packages = array();
+	foreach ( glob( gutenberg_dir_path() . 'build/esm/*.js' ) as $path ) {
+		$package_name = substr( basename( $path ), 0, -3 );
+		$packages[]   = 'wp-' . $package_name;
+	}
+	$vendors                 = array( 'lodash', 'react', 'react-dom', 'moment', 'moment-timezone' );
+	$export_default_packages = array(
+		'api-fetch',
+		'deprecated',
+		'dom-ready',
+		'redux-routine',
+		'token-list',
+		'server-side-render',
+		'shortcode',
+		'warning',
+	);
+
+	if ( in_array( $handle, $packages, true ) ) {
+		$package_name = substr( $handle, 3 );
+		$import       = in_array( $package_name, $export_default_packages, true ) ? 'mod' : '* as mod';
+		// Replace all scripts with a module script to defer them.
+		$tag = str_replace( '<script ', '<script type="module" ', $tag );
+
+		// Replace the content script with an esmodule import.
+		$tag          = str_replace(
+			sprintf( "<script type=\"module\" src='%s' id='%s-js'></script>", $src, esc_attr( $handle ) ),
+			sprintf(
+				'<script id="script-%s" type="module">import %s from "@wordpress/%s"; window.wp = window.wp || {}; window.wp.%s = mod;</script>',
+				$package_name,
+				$import,
+				$package_name,
+				gutenberg_to_camel_case( $package_name )
+			),
+			$tag
+		);
+	} elseif ( in_array( $handle, $vendors, true ) ) {
+		$tag = sprintf(
+			'<script id="script-%s" type="module">import * as mod from "%s"; window.%s = mod;</script>',
+			$handle,
+			$handle,
+			gutenberg_to_camel_case( $handle )
+		);
+	} elseif ( false === strpos( $handle, 'thickbox' ) && false === strpos( $handle, 'tinymce' ) && false === strpos( $handle, 'clipboard' ) && false === strpos( $handle, 'jquery' ) && false === strpos( $handle, 'hoverintent' ) && false === strpos( $handle, 'moxie' ) ) {
```

Comment:
There are scripts across WP code base that inject `<script>` tags without relying on `script_loader_tag` filter, meaning if they have dependencies (like tinymce), that dependency shouldn't be a script either, that's why I added all these exceptions. There's probably a better way here.